### PR TITLE
optimize_for_mobile: bring packed params to root module

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42041 optimize_for_mobile: bring packed params to root module**
* #42461 fix asan failure for fold_conv_bn when conv is quantized

Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

### CODE REVIEW

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (vkuzo) [inline torch/csrc/jit/passes/hoist_conv_packed_params.cpp:109]: would love some tips on best way to do this

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp
Comment (kimishpatel) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: I am assuming this requires inlining? If so then you might want to run freeze_module once again? Since other pass might be disabled?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Looks pretty good. Left one comment.

File: test/test_mobile_optimizer.py
Comment (jerryzh168) [inline test/test_mobile_optimizer.py:279]: nit: can you put these two test cases in a loop?

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/hoist_conv_packed_params.cpp:55]: we have some helper function in quantization to do this: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16getInvokedModuleERNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE, feel free to use them

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/hoist_conv_packed_params.cpp:69]: all code above can be achieved by getInvokedModule and getModuleAccessPath I think

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/hoist_conv_packed_params.cpp:109]: please checkout: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16is_conv2d_moduleERKNS0_5MatchERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS0_5ValueESt4hashISA_ESt8equal_11899141

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/hoist_conv_packed_params.cpp:55]: oh this one: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit19getInvokedModuleOptERKNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE since maybe we will fail to get module

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/hoist_conv_packed_params.cpp:77]: if the name already exists maybe we can create a new one? e.g. https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#937

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/hoist_conv_packed_params.cpp:66]: lint

File: test/test_mobile_optimizer.py
Comment (jerryzh168) [inline test/test_mobile_optimizer.py:268]: also can we check the access to conv module is removed?

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: yeah we may need to freeze again to remove the conv module itself I think

File: test/test_mobile_optimizer.py
Comment (jerryzh168) [reply test/test_mobile_optimizer.py:279]: and conv module attribute is removed

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (vkuzo) [reply torch/csrc/jit/passes/hoist_conv_packed_params.cpp:66]: I would imagine if we hit a duplicate it would be more likely that this code has a bug, compared to the user having something named `_jit_pass_hoist_conv_packed_params.{foo}.{bar}.{conv_foobar}` before this pass.

File: test/test_mobile_optimizer.py
Comment (vkuzo) [reply test/test_mobile_optimizer.py:266]: sure, will add

Comment (vkuzo) [reply test/test_mobile_optimizer.py:266]: actually, sorry, what do you mean specifically?  The original access is 
```
__torch__.torch.nn.quantized.modules.conv.___torch_mangle_13.Conv2d = prim::GetAttr[name="conv1"](%self)
```

I'm checking for the absence of a suffix of that (for readability).  Happy to adjust if something else is better

Comment (jerryzh168) [reply test/test_mobile_optimizer.py:266]: I mean check the conv attribute is removed from the parent module

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/hoist_conv_packed_params.cpp:66]: if `.` can appear in attribute names then we might hit duplicate in extreme cases, although it is very unlikely
e.g. consider {foo}.{bar} is also an attribute itself

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: I also meant before. Doesnt this pass require inlining of the graph as well? Because right now inlining is done by freeze_module, and if you have INSERT_FOLD_PREPACK_OPS blocklisted, it wont do that. So if inlining is required I suggest you call Inline explicitly before calling `HoistConvPackedParams`

File: test/test_mobile_optimizer.py
Comment (vkuzo) [reply test/test_mobile_optimizer.py:266]: makes sense.  Any tips on the best way to do that with `FileCheck`?

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (vkuzo) [reply torch/csrc/jit/passes/hoist_conv_packed_params.cpp:66]: good point, thx, will update

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: yeah, I have the same question, we need to make sure what is required before running this pass. I think inline is required in this case. Also actually I'm not sure if it is OK to freeze module for user if it is not frozen previously.

File: test/test_mobile_optimizer.py
Comment (jerryzh168) [reply test/test_mobile_optimizer.py:266]: FileCheck is just checking the strings. for this one we need to check the module itself with something like
`self.assertFalse(hasattr(module, 'conv1'))`

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: @kimishpatel  In general I feel this is an optimization only valid when freeze is enabled, is there a way to enforce that in `optimization_blocklist`?

Comment (vkuzo) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: sounds good.  We can throw an assert here for `INSERT_FOLD_PREPACK_OPS` being enabled, unless there is a better way.

Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:304]: > @kimishpatel In general I feel this is an optimization only valid when freeze is enabled. why is freezing not an optimization option actually?

@jerryzh168, I think freezing was never thought of as an optimization think when I did this. Your point is fair, maybe we can add that.

@vkuzo, I would say instead of throwing an assert, we just add freezing before this. I understand it may mean we are calling freezing multiple times, but that should not be an issue. Plus that makes this pass self contained without any dependency on having to run another pass.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp
Comment (vkuzo) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:302]: I'm seeing optimize_for_mobile test fail on this line (https://gist.github.com/vkuzo/0e0562fd46a10b9ed8f2f6e8eefca349), just checking that this is expected and we can update the test to match?  seems like that test case currently assumes the BN::forward call survives

Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:302]: No I dont think so.  BNTestModule has BN and Conv2d. The test where it fails, it has disabled prepack folding. If that is disabled there is no inlining, via freeze_module. So we should still have Conv2d::forward. I think that is what it is expecting. But since you have freeze_module after, and you probably will add freeze_module before, inlining will happen and there wont be any forward calls. So yeah we will have to fix that test somehow. Maybe best to remove that portion of the test.

**Review by jerryzh168 — APPROVED** [review]
LGTM, please fix the failing tests.

**Comment by vkuzo** [comment]
we should wait for https://github.com/pytorch/pytorch/issues/42462 to be resolved before landing this

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (raghuramank100) [inline torch/csrc/jit/passes/hoist_conv_packed_params.cpp:39]: Can we also add a pass for linear? Is there value in doing this for linear modules too?

File: test/test_mobile_optimizer.py
Comment (raghuramank100) [inline test/test_mobile_optimizer.py:249]: Is the understanding that one cannot serialize/deserialize a model after optimize_for_mobile is called?

File: torch/csrc/jit/passes/hoist_conv_packed_params.cpp
Comment (vkuzo) [reply torch/csrc/jit/passes/hoist_conv_packed_params.cpp:39]: I don't see why not, although I'd defer to the mobile team on the priority.

**Review by kimishpatel — APPROVED** [review]
Looks good. Thanks.

**Comment by vkuzo** [comment]
resubmitted as https://github.com/pytorch/pytorch/pull/42740 due to a ghstack snafu

**Pull request closed without merging**
